src/lbann_library.cpp

```diff
@@ -0,0 +1,320 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/lbann_library.hpp"
+#include <dirent.h>
+#include <cstdlib>
+
+namespace lbann {
+
+model * __attribute__((used)) build_model_from_prototext(int argc, char **argv,
+                                                         lbann_data::LbannPB &pb,
+                                                         lbann_comm *comm,
+                                                         bool first_model) {
```

Comment:
It's not a good sign that we have to pass `first_model` in, since it shows this routine has complicated side effects:
- The first time we setup a model, we call `comm->split_models()`. It would be better to do this when we setup the communicator.
- The first time we setup a model, we do a hack to share data readers amongst models. In particular, we load the data reader into a global `options` object. This is not nice.

Reply:
> The first time we setup a model, we do a hack to share data readers amongst models. In particular, we load the data reader into a global options object. This is not nice

mea culpa. This was my hack to reduce memory usage. It's  currently not used, but  may be in the future -- but (as of now) only for  cycgan.  I'm OK with deleting this.

Reply:
Heh, I might be too harsh. Hacks are fine, but we should periodically clear them out before they bite us.

Reply:
agreed.

Reply:
I agree that the first_model is hack that we want to find a more elegant way to support.  At this point I really wanted to consolidate some of the code so that we can put in the inference front end for immediate use.  After the current deadline we can figure out how to refactor this more cleanly.  This will also take into account the pending conversation about multi-model training methods.